### ISSUE TITLE
Make resolver TTL configurable

### ISSUE THREAD

**Issue by nschoe** [docker/docker#30029]
I ran into a problem where I had an nginx that acted as front-end and was connected to a lot of docker networks to dispatch queries.
All is linked with `--net-alias` so I can rely on the internal resolver instead of IP.

Turns out, sometimes the containers need to be restarted and they change IP, but the nginx catches the resolved query for as long as the resolver's TTL.

There is a workaround to force nginx to use a different `valid=` time, but it would be good if it was configurable.

@rels on #docker pinpointed the ligne [here](https://github.com/docker/libnetwork/blob/master/resolver.go#L66)

**Comment by YouriT** [comment]
Indeed I don't see any setter on the TTL, only getters.

We discussed also about `--dns-opt` but it doesn't seem to solve the issue here as it only refers to `/etc/resolv.conf` options.

Swarm DNS round-robin seems also to be affected by this issue. As the TTL is also 600:
```
# dig +nocmd +noall +answer test
test.			600	IN	A	10.0.0.17
test.			600	IN	A	10.0.0.19
```

Swarm VIP shouldn't have this issue I believe.

**Comment by thaJeztah** [comment]
@nschoe are you sure your issue is related to TTL? IIRC, the community (free) edition of NGINX is only resolves domains when it starts up, and never updates those results; https://tenzer.dk/nginx-with-dynamic-upstreams/

**Comment by YouriT** [comment]
@thaJeztah that's true but it is a workaround. If you have a fully compliant service that respects TTL, you would still get the same issue no?

Why is the TTL fixed? And why can't it be changed?

**Comment by sanimej** [comment]
@YouriT Docker DNS server right now doesn't have a UX to set custom TTL values. Its something we can consider adding if there are more use cases for configurable settings.

**Comment by nschoe** [comment]
@thaJeztah I'm pretty sure, here is what I tested:

- I set up my stack (several containers in their own Docker Network and --net-alias, a front-end nginx, connected to every Docker network, and it resolves based on the --net-alias)
- When it works, I force delete / recreate one container, which changes IP (but same --net-alias)
- When requesting the site and watching nginx logs, I clearly see the "fail, upstream <OLD_IP_HERE> not responding" message. And only after some time, does the URL works again

The workaroun is to use `valid=30s` (for instance) [here](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver) in nginx's `resolver` option.
The doc confirms that when none are present (my case), the `valid=` takes the TTL value.

**Comment by nschoe** [comment]
@sanimej yeah I think it should be good.

For information I was not criticizing at all: needing to tune the Docker Internal DNS resolver becomes a pretty advanced usage, and my network stack is indeed fairly complex.
So I'm not "blaming" docker for not making this tunable :-)

I'm just saying that "this is it", people (me ^^) begins to need it :-)

A quite easy way would be to pass this parameter in the docker daemon, but what would be even better is if we could set it up on the fly with the docker CLI. I'm not sure how hard this will be to implement.

I'm sad I haven started learning Go yet as I'd love to make a PR :-)

To make things a bit clearer why I need this: basically I have one container + Docker network stack for each client I have, and a front-end nginx connected to all of these Docker network stacks. The front-end nginx provides a single point of entry for my API calls, which then relies upon Docker Internal DNS solver and --net-alias to route API queries.

When a client decides to update his system, an image is rebuilt, the containers in his stack are destroys and recreated based on the new image.
This necessarily changes the container's IP, and make the new site unreachable for up to 600 seconds (it is of course rare that the container update happens precisely after a new TTL, but still, it's possible).

I mitigated this by displaying a "loading gear" and displaying some "updating in progress" but this is frustrating because the image rebuild and container creation is very quick, I'm actually limited by the TTL :-)

I will try changing the nginx' `valid=` setting in the meantime :-)

Cheers!

**Comment by dlozlla** [comment]
Being able to configure this TTL would be much appreciated, especially after seeing problems with VIPs (https://github.com/moby/moby/issues/32195)

**Comment by dimaqq** [comment]
Poke 🙏🏿 
I would like [a lot] faster TTL than 600 too
in my case, it's for local testing of something that is ultimately deployed with other tools

**Comment by ncresswell** [comment]
Has any progress been made on this? Having DNS TTL set to 600 and not configurable really does become a problem when tasks fail often.
